<h1 style="text-align: center;"><a title="Data Science-AIMS-Cmr-2021-22">Chapter 2: Data Reading Methods </h1>

**Instructor:** 

* Rockefeller

# Introduction

Before diving deep into data manipulation and analysis, mastering the art of data ingestion using pandas is pivotal. This chapter will guide you through the essentials of reading data from various sources, ensuring a robust foundation for your data exploration journey.

**Learning Objectives:**

* Introducing different files reading methods
* Developing the ability to query dataframes to answer practical questions
* Updating dataframes based on useful insights

# Reading Methods

Python has become a beloved language for text and file munging due to its simple syntax for
interacting with files, intuitive data structures, and convenient features. Pandas features a number of functions for reading tabular data as a DataFrame object. Let's import it

In [1]:
import pandas as pd

A plethore of  reading method is available in the pandas library.  You can have access to it by typing the `tab` key after typing `pd.`:


>#### <font color=#800080>Q:</font> <a class="anchor" id="Task-1"></a>

* How many of them could you see and at first glance, what could they possibly mean to you?

For sake of the context , we will just focus on the 3 main reading methods

* $\Large pd.read\_csv$


* $\Large pd.read\_excel$


* $\Large pd.read\_clipboard$

### The `pandas.read_csv` method 

Reading a csv file requires passing as argument the path to the actual file. For simplicity,  the file should be saved in the same folder as the notebook you are running but you could also specify the path to it. See below.

Here is the pattern to follow:  

`dataframe_name  =  pd.read_csv( "path/to/the/file.csv" )`

In [5]:
import pandas as pd

In [7]:
df_ebola_west_africa = pd.read_csv("/home/students-aimssn/Documents/Data_Prep/Module_02/data/ebola_WA.csv")

In [8]:
df_ebola_west_africa

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
0      1/5/2015  289        2776.0            NaN            10030.0   
1      1/4/2015  288        2775.0            NaN             9780.0   
2      1/3/2015  287        2769.0         8166.0             9722.0   
3      1/2/2015  286           NaN         8157.0                NaN   
4    12/31/2014  284        2730.0         8115.0             9633.0   
..          ...  ...           ...            ...                ...   
117   3/27/2014    5         103.0            8.0                6.0   
118   3/26/2014    4          86.0            NaN                NaN   
119   3/25/2014    3          86.0            NaN                NaN   
120   3/24/2014    2          86.0            NaN                NaN   
121   3/22/2014    0          49.0            NaN                NaN   

     Cases_Nigeria  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
..             ...  
117            NaN  
118            NaN  
119            NaN  
120            NaN  
121            NaN  

[122 rows x 6 columns]

># <font color=#800080>Task 2:</font> <a class="anchor" id="Task-1"></a>


Dr. Amara stood at the front of the quiet conference room on a gloomy afternoon. 

The projector screen behind him displayed the year 2014, a year that everyone in the room knew was significant for West Africa.


1. In 2014, the Ebola crisis broke out in some of our neighboring countries. But before we talk about the details, can anyone explain the difference between an outbreak and a pandemic?



2. After a pause, Dr. Amara pointed to a map of Africa on the screen and said, **"Ebola is not a new virus, but it spread very quickly in 2014.** Does anyone know where it came from and why it spread so fast that year?



3. After a short discussion, Dr. Amara led the group to the day's task. She said, **"On your computers, you will find a folder called `data`. Inside that folder is a file called `ebola_WA.csv`**. This file contains information about how many Ebola cases were reported in **Nigeria**, **Guinea**, and **Sierra Leone** over time. Please take a moment to make sure that the file is there and then load it using the standard methods that we have learned.


4. A few minutes later, as the data appeared on the screens, Dr. Amara asked, "What strange things or patterns do you see in the data? Let's talk about them."

### Outbreak affects a particular place while Pandemic affects the whole world

># <font color=#800080>Task 3:</font> <a class="anchor" id="Task-1"></a>


Dr. Amara paused and spoke with understanding. She said, **"Data analysts are often not involved in collecting data. This means we sometimes work with datasets that have unfamiliar structures or entries."**

1. Dr. Amara switched to a new slide and said, **"I have a new set of historical Ebola data from our colleagues in Central Africa**. Please take a look at the data in the file called `'ebola_CA.csv'` and share your first thoughts."


2. Dr. Amara saw that people were starting to look at the new data. She said, **"Once you've looked through the data, use the same method we learned earlier to put the data into a table. Then, tell us what you see."**

In [2]:
ebola_data_CA =  pd.read_csv('data/ebola_CA.csv', sep='-')
ebola_data_CA.head()

Unnamed: 0        Date  Day  Cases_Cameroon  Cases_Gabon  Cases_CAR
0           0    1/5/2015  289            50.0         25.0      242.0
1           1    1/4/2015  288            69.0          NaN        NaN
2           2    1/3/2015  287            45.0         93.0      285.0
3           3    1/2/2015  286           395.0         13.0      157.0
4           4  12/31/2014  284           300.0         69.0      250.0

You might have guessed it. CSV stands for comma separated values. Other files similar file extensions are also used. See below:

* `","` for the csv file
* `"\t"` for a tsv file 
* or any other character that separates the different variables for an observation

As example of the  `.tsv` file, see the gapminder dataset below. Let's read this dataset where variables are separated by a tabulation key.

In [3]:
import pandas as pd

In [4]:
world_gapminder = pd.read_csv('data/gapminder.tsv' , sep = "\t")
world_gapminder.head(5)

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106

###  The `pandas.read_excel` method 

Excel files are also quite common files for saving data. The XLS file extension is used for files saved as Microsoft Excel worksheets. Reading an excel file requires the same process as reading a csv file. The only difference lies in specifying the sheet you want to get access to.

The dataset below comes from the mining industry. It contains mining locations in some African countries both at the region and at the district level. See here a link to an interesting website where you can download the shapefiles of such regions.  https://gadm.org/formats.html 

The code below accesses an excel sheet nameed `ds_salaries`and the sheet `salary` .

In [5]:
df_salaries = pd.read_excel('data/pro_salaries.xlsx', sheet_name = 'salary')

df_salaries.head(5)

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

**Just from looking at the dataset, at first glance, can you guess the meaning of some of the columns?**

### The `pandas.read_clipboard` method 

The `pandas.read_clipboard` method is a convenient tool that reads the contents of your clipboard directly into a pandas DataFrame. This method can be quite useful when you want to quickly grab tables or data that you've copied from websites, spreadsheets, or other sources without having to save them as files first.


In [24]:
# lets practice some.  
# Take a look at this website https://www.timeanddate.com/

    #--->  Navigate to Sun, Moon and Space
    # --->  Click on Sun Calculator  
    # --->  Enter the name of a city of your choice in the search box and click on it.


import pandas as pd
my_city_df = pd.read_clipboard()
my_city_df

PyperclipException: 
    Pyperclip could not find a copy/paste mechanism for your system.
    For more information, please visit
    https://pyperclip.readthedocs.io/en/latest/index.html#not-implemented-error
    

There are several other pandas reading methods. Depending on the files you are dealing with, you might have to choose the convenient method. Other popular methods include `pd.read_json` (for `json` files) , `pd.read_html` (for `html` files) , `pd.read_`

The principle stays the same, which is reading from external sources and casting it into a pandas dataframe for facilitating the analysis. 

##  Shape and datatype

In [26]:
ebola_wa = pd.read_csv('data/ebola_WA.csv')

To get the first n rows and the last n rows

In [29]:
ebola_wa.head(4)

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
0  1/5/2015  289        2776.0            NaN            10030.0   
1  1/4/2015  288        2775.0            NaN             9780.0   
2  1/3/2015  287        2769.0         8166.0             9722.0   
3  1/2/2015  286           NaN         8157.0                NaN   

   Cases_Nigeria  
0            NaN  
1            NaN  
2            NaN  
3            NaN

In [8]:
ebola_wa.tail(5)

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
117  3/27/2014    5         103.0            8.0                6.0   
118  3/26/2014    4          86.0            NaN                NaN   
119  3/25/2014    3          86.0            NaN                NaN   
120  3/24/2014    2          86.0            NaN                NaN   
121  3/22/2014    0          49.0            NaN                NaN   

     Cases_Nigeria  
117            NaN  
118            NaN  
119            NaN  
120            NaN  
121            NaN

A random sample of 3 rows 

In [31]:
ebola_wa.sample(3)

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
42  10/7/2014  199        1350.0         4076.0                NaN   
57   9/5/2014  167         812.0         1871.0             1261.0   
58  8/31/2014  162         771.0         1698.0             1216.0   

    Cases_Nigeria  
42            NaN  
57           22.0  
58           21.0

To check the shape of the dataframe (rows and columns)

In [32]:
ebola_wa.shape

(122, 6)

To check the datatype of each of the columns

In [11]:
ebola_wa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122 entries, 0 to 121
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Date               122 non-null    object 
 1   Day                122 non-null    int64  
 2   Cases_Guinea       93 non-null     float64
 3   Cases_Liberia      83 non-null     float64
 4   Cases_SierraLeone  87 non-null     float64
 5   Cases_Nigeria      38 non-null     float64
dtypes: float64(4), int64(1), object(1)
memory usage: 5.8+ KB


* We can notice that the dataset contains 122 entries with their index ranging from 0 to 121

* The columns  **Cases_Guinea, Cases_Liberia, Cases_SierraLeone** and **Cases_Nigeria** are **float64** type.
Which means they contain floating points.

* The ***Day** column is of **int64** type

* Some of the columns contains missing values:
        122-93 = 29 for the Cases_Guinea
        122-83 = 39 for the Cases_Liberia
        122-87 = 35 for the Cases_SierraLeone
        122-38 = 84 for the Cases_Nigeria

To check the list of columns

In [33]:
ebola_wa.columns

Index(['Date', 'Day', 'Cases_Guinea', 'Cases_Liberia', 'Cases_SierraLeone',
       'Cases_Nigeria'],
      dtype='object')

Let's consider another dataset

In [13]:
world_gapminder = pd.read_csv('data/gapminder.tsv' , sep='\t')
world_gapminder.head(5)

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106

To check the unique values inside a column

In [34]:
world_gapminder['continent'].unique()

array(['Asia', 'Europe', 'Africa', 'Americas', 'Oceania'], dtype=object)

In [35]:
world_gapminder['country'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
       'Australia', 'Austria', 'Bahrain', 'Bangladesh', 'Belgium',
       'Benin', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Czech Republic',
       'Denmark', 'Djibouti', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Ethiopia',
       'Finland', 'France', 'Gabon', 'Gambia', 'Germany', 'Ghana',
       'Greece', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Haiti',
       'Honduras', 'Hong Kong, China', 'Hungary', 'Iceland', 'India',
       'Indonesia', 'Iran', 'Iraq', 'Ireland', 'Israel', 'Italy',
       'Jamaica', 'Japan', 'Jordan', 'Kenya', 'Korea, Dem. Rep.',
       'Korea, Rep.', 'Kuwait', 'Leba

In [ ]:
world_gapminder['country'].nunique() # To know total number of country or item

142

In [39]:
world_gapminder['year'].unique()

array([1952, 1957, 1962, 1967, 1972, 1977, 1982, 1987, 1992, 1997, 2002,
       2007])

In [40]:
world_gapminder['lifeExp'].unique()

array([28.801, 30.332, 31.997, ..., 46.809, 39.989, 43.487], shape=(1626,))

In [41]:
world_gapminder['pop'].unique()

array([ 8425333,  9240934, 10267083, ..., 11404948, 11926563, 12311143],
      shape=(1704,))

In [42]:
world_gapminder['gdpPercap'].unique()

array([779.4453145, 820.8530296, 853.10071  , ..., 792.4499603,
       672.0386227, 469.7092981], shape=(1704,))

In [ ]:
world_gapminder['continent'].value_counts() # number of occurences of continent 

continent
Africa      624
Asia        396
Europe      360
Americas    300
Oceania      24
Name: count, dtype: int64

To get a subset of the data, let's query the dataframe and get some insights

># <font color=#800080>Task 4:</font> <a class="anchor" id="Task-1"></a>


1. Maria, a public health student, was sitting in a busy cafe in Luanda. She turned to her colleague, João, and asked him, **"Do you remember when Angola's life expectancy was above 40 years old?** For a country, what do you think large value of life expectancy can be an indicator of? Find those years in the dataset and report their corresponding population level.



2. Dr. Kofi and Dr. Amina, two economics professors, were sitting under a large acacia tree, enjoying their afternoon tea. Dr. Kofi looked nostalgic as he asked Dr. Amina, "Which African country do you think had the highest GDP in the 80s?"  Could you guess the name of that country before computing it?? After computing it, what are your thoughts about that?

># <font color=#800080>Task 5:</font> <a class="anchor" id="Task-1"></a>


Look at the gapminder dataset and answer the following questions:

1. **Southern European GDP in the 2000s:**

At a financial conference in Johannesburg, people were talking about how the economy is changing. Sofia, a delegate from Greece, turned to Kwame, a well-known economic analyst from Ghana, and asked him, **"Kwame, do you know which Southern European countries had the highest and lowest GDP in the 2000s?"** 
Hint: Use the Southern European countries listed below.

2. **African Life Expectancy in the 1970s:**

At a health and development conference in Addis Ababa, people were talking about how healthcare has changed as compared to in the past. Fatima, a public health student from Algeria, went up to Zola, a respected demographic analyst from Tanzania, and asked her, **"Zola, can you tell me which African countries had the highest and lowest life expectancy in the 1970s? What do you think caused this?"**


3. **Asian vs. European GDP in the 90s:**


During an international economics conference in Dakar, people were talking about which countries used to be the richest. Ravi, a researcher from India, curiously asked Chijioke, a skilled economic analyst from Cameroon, **"Chijioke, how many Asian countries do you think had a higher GDP than the average European country in the 1990s?"**

In [ ]:
SE_countries = ['Albania', 'Andorra', 'Bosnia and Herzegovina', 'Croatia', 'Gibraltar', 
                   'Greece', 'Italy', 'Kosovo', 'Malta', 'Montenegro', 'North Macedonia', 
                   'Portugal', 'San Marino', 'Serbia', 'Slovenia', 'Spain', 'Vatican City']